lang/wasm.cc

```diff
@@ -0,0 +1,279 @@
+/*
+ * Copyright (C) 2021-present ScyllaDB
+ */
+
+/*
+ * This file is part of Scylla.
+ *
+ * Scylla is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Scylla is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifdef SCYLLA_ENABLE_WASMTIME
+
+#include "wasm.hh"
+#include "concrete_types.hh"
+#include "utils/utf8.hh"
+#include "utils/ascii.hh"
+#include "utils/date.h"
+#include "db/config.hh"
+#include <seastar/core/coroutine.hh>
+#include <seastar/util/defer.hh>
+#include "seastarx.hh"
+
+static logging::logger wasm_logger("wasm");
+
+namespace wasm {
+
+context::context(wasm::engine* engine_ptr, std::string name) : engine_ptr(engine_ptr), store(engine_ptr->get()), function_name(name) {
+    auto added = store.context().add_fuel(engine_ptr->initial_fuel_amount());
+    if (!added) {
+        throw wasm::exception(added.err().message());
+    }
+}
+
+void compile(context& ctx, const std::vector<sstring>& arg_names, std::string script) {
+    wasm_logger.debug("Compiling script {}", script);
+    auto module = wasmtime::Module::compile(ctx.engine_ptr->get(), script);
+    if (!module) {
+        throw wasm::exception(format("Compilation failed: {}", module.err().message()));
+    }
+    ctx.module = module.unwrap();
+    auto instance = wasmtime::Instance::create(ctx.store, *ctx.module, {});
+    if (!instance) {
+      throw wasm::exception(format("Creating a wasm runtime instance failed: {}",instance.err().message()));
+    }
+    ctx.instance = instance.unwrap();
+    auto function_obj = ctx.instance->get(ctx.store, ctx.function_name);
+    if (!function_obj) {
+        throw wasm::exception(format("Function {} was not found in given wasm source code", ctx.function_name));
+    }
+    wasmtime::Func* fun = std::get_if<wasmtime::Func>(&*function_obj);
+    if (!fun) {
+        throw wasm::exception(format("Exported object {} is not a function", ctx.function_name));
+    }
+    ctx.func = std::move(*fun);
```

Comment:
Do we compile for every invocation? Or just once?

Somewhere we need to copy the contents of the data section into the memory, just before executing, so initialized global variables work.

Reply:
We compile once on function creation and each time its schema is propagated (since we keep the source code in schema). During compilation, we produce not only code but create an instance (ref from js: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WebAssembly/Instance) , which is a sandbox-like independent unit of execution which contains all the functions and state. Now that I wrote it, I think that the current implementation would infinitely grow the memory until it bursts - because each call to a UDF grows the memory, and then it's never reclaimed. That would need to change -- by resetting the store (https://docs.wasmtime.dev/api/wasmtime/struct.Store.html) to an initial state.

Reply:
(or just by recreating the instance for each call, which might be expensive-ish)